types/document.d.ts

```diff
@@ -132,7 +132,7 @@ declare module 'mongoose' {
     getChanges(): UpdateQuery<this>;
 
     /** The string version of this documents _id. */
-    id?: any;
+    id?: string;
```

Comment:
This is a hard no. This would break anyone's code who uses `extends Document` with an `id` field that isn't a string.

Reply:
> Okay, here is a simple test that shows one issue, though it's pretty contrived:
> 
> ```ts
> interface ThingType {
>   name: string;
> }
> const thingSchema = new Schema<ThingType>({
>     name: Schema.Types.String
> });
> const ThingModel = model('Thing', thingSchema);
> type ThingDocument = HydratedDocument<ThingType>;
> // or
> type ThingyDocument = InstanceType<typeof ThingModel>;
> 
> function AcceptLeanDoc(val: LeanDocument<ThingDocument>) {
>     // This should fail but it thinks the type is valid
>     val.populate('');
> }
> ```
> 
> <img alt="image" width="503" src="https://user-images.githubusercontent.com/224334/166824987-05a0ad9d-e2a5-49c5-99f0-757b3d4e1753.png">
> 
> Your type isn't actually constructing a lean type, it's starting with a type that is already lean -- which is great when you can do that, but there are cases where you can't or it's impractical. At best you have an extra type you have to pass around everywhere.

I think this will be so instead:
```
async function _11762() {
  interface ThingType {
    name: string;
  }
  const thingSchema = new Schema<ThingType>({
    name: Schema.Types.String
  });
  const ThingModel = model('Thing', thingSchema);

  function AcceptLeanDoc(val: LeanDocument<ThingType>) {
    // This should fail but it thinks the type is valid
    val.populate('');
  }
}
```

Reply:
This would break anyone's code who uses extends Document with an id field that isn't a string.
But then we have to update docs because it says it will be string.
Or users can omit id type before in type overriding case.

Reply:
You can do new Schema({ id: Number }), that's perfectly valid in Mongoose. Mongoose just overrides the default id virtual.